src/main/java/org/jenkinsci/plugins/workflow/support/pickles/serialization/RiverWriter.java

```diff
@@ -169,26 +164,72 @@ public ObjectOutput getObjectOutput() {
 
     public void close() throws IOException {
         marshaller.finish();
-        int ephemeralsOffset = dout.size();
+        int ephemeralsOffset = (int)channel.position();
 
         // write the ephemerals stream
         pickling = false;
-        marshaller.start(Marshalling.createByteOutput(dout));
+        marshaller.start(new FileChannelOutput(channel));
         marshaller.writeObject(pickles);
         marshaller.finish();
-        dout.close();
 
         // back fill the offset to the ephemerals stream
-        RandomAccessFile raf = new RandomAccessFile(file, "rw");
         try {
-            raf.seek(ephemeralsBackptr);
-            raf.writeInt(ephemeralsOffset);
+            channel.position(EPHEMERALS_BACKPTR);
+            ByteBuffer ephemeralsPtrBuffer = ByteBuffer.allocate(4).putInt(ephemeralsOffset);
+            ephemeralsPtrBuffer.flip();
+            channel.write(ephemeralsPtrBuffer);
+            channel.force(true);
```

Comment:
I guess metadata is not really important here, so maybe we should use `force(false)` (`fdatasync`) for potentially slightly better performance?